### PR TITLE
Feature/43342 ps report data sync

### PR DESCRIPTION
**Pull request by jon-shipley**
Persist pupil device data to SQL Server

### CODE REVIEW

**Review by GuyHarwood — APPROVED** [review]
LGSF.  Has merge conflicts to resolve

File: admin/data/sql/migrations/20201015195026.do.update-userdevice-table.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+        SELECT *
+          FROM sys.columns
+         WHERE object_ID = object_id('mtc_results.userDevice')
+           AND col_name(object_ID, column_Id) = 'ident')
+    BEGIN
+        ALTER TABLE [mtc_results].[userDevice]
+            ADD [ident] NVARCHAR(100);
+    END
+
+
```
Comment (GuyHarwood) [inline admin/data/sql/migrations/20201015195026.do.update-userdevice-table.sql:11]: 💯

File: bin/run-all-unit-tests.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
```
Comment (GuyHarwood) [inline bin/run-all-unit-tests.sh:1]: useful 🎖️

File: pupil-spa/src/app/services/device/device.service.spec.ts [hunk 3]
```diff
@@ -37,9 +41,20 @@ describe('DeviceService', () => {
...
+    // test setup
+    service.deleteDeviceCookie();
+    service.setupDeviceCookie();
+
+    // exec
     await service.capture();
+
+    // tests
     const deviceInfo = storageService.getDeviceData();
     expect(deviceInfo).toBeTruthy();
```
Comment (GuyHarwood) [inline pupil-spa/src/app/services/device/device.service.spec.ts:53]: when you pull in the tslib lint update PR, these type of assertions will fail linting.

please use `toBeInstanceOf`, `toBeDefined` and `toBe(true)` instead

File: pupil-spa/src/app/services/device/device.service.spec.ts [hunk 4]
```diff
@@ -37,9 +41,20 @@ describe('DeviceService', () => {
...
+    expect(deviceInfo.deviceId.length).toBe(36); // uuid as string
```
Comment (GuyHarwood) [inline pupil-spa/src/app/services/device/device.service.spec.ts:54]: new linter will fail this in favor of `toHaveLength(36)`

File: tslib/src/functions-throttled/sync-results-to-sql/sync-results.data.service.ts [hunk 5]
```diff
@@ -93,12 +96,235 @@ export class SyncResultsDataService implements ISyncResultsDataService {
...
+                @userAgentLookup_id,
+                @ident);
+
+        SET @userDeviceId = (SELECT SCOPE_IDENTITY());
+
+        UPDATE mtc_results.checkResult
+           SET userDevice_id = @userDeviceId
+         WHERE id = @checkResultId;
+    `
+    // tslint:enable:no-trailing-whitespace
```
Comment (GuyHarwood) [inline tslib/src/functions-throttled/sync-results-to-sql/sync-results.data.service.ts:324]: please also review this directive when you pull in the updated linter

File: tslib/src/functions-throttled/sync-results-to-sql/sync-results.sync.spec.ts [hunk 6]
```diff
@@ -27,6 +27,7 @@ describe('SyncResultsService', () => {
     ;(mockSyncResultsDataService.prepareCheckResult as jest.Mock).mockReturnValueOnce(mockTransaction)
     ;(mockSyncResultsDataService.prepareAnswers as jest.Mock).mockReturnValueOnce(mockTransaction)
     ;(mockSyncResultsDataService.prepareEvents as jest.Mock).mockReturnValueOnce(mockTransaction)
+    ;(mockSyncResultsDataService.prepareDeviceData as jest.Mock).mockReturnValueOnce(mockTransaction)
```
Comment (GuyHarwood) [inline tslib/src/functions-throttled/sync-results-to-sql/sync-results.sync.spec.ts:30]: why the semi-colon prefix?

**Review by GuyHarwood — CHANGES_REQUESTED** [review]
looks great, especially the utility stuff you added.

Just a few changes to make in tests, as the new linter rules will reject some of the apis/styles you are using

File: pupil-spa/src/app/services/device/device.service.spec.ts (hunk 3 above)
Comment (jon-shipley) [reply pupil-spa/src/app/services/device/device.service.spec.ts:53]: Or possibly not yet?  Master was merged in yesterday, and that included the eslint changes.   This is the SPA, and I see a couple of  issues:

1. Neither `yarn test` or `yarn test-single` do a lint run
2. `yarn lint` actually calls `ng lint` which passes

The SPA does not seem to have been updated to run `eslint` yet.

File: pupil-spa/src/app/services/device/device.service.spec.ts (hunk 4 above)
Comment (jon-shipley) [reply pupil-spa/src/app/services/device/device.service.spec.ts:54]: That is not a valid `jasmine` matcher for version 2.8.
https://jasmine.github.io/api/2.8/matchers.html

Are you thinking of `jest`? 😄

File: tslib/src/functions-throttled/sync-results-to-sql/sync-results.data.service.ts (hunk 5 above)
Comment (jon-shipley) [reply tslib/src/functions-throttled/sync-results-to-sql/sync-results.data.service.ts:324]: The new linter is in!

File: tslib/src/functions-throttled/sync-results-to-sql/sync-results.sync.spec.ts (hunk 6 above)
Comment (jon-shipley) [reply tslib/src/functions-throttled/sync-results-to-sql/sync-results.sync.spec.ts:30]: It prevents the js engine from thinking the brackets are function call parameters to the preceding line.

File: pupil-spa/src/app/services/device/device.service.spec.ts (hunk 3 above)
Comment (jon-shipley) [reply pupil-spa/src/app/services/device/device.service.spec.ts]: I've changed it to `toBeDefined()` 😄

File: pupil-spa/src/app/services/device/device.service.spec.ts (hunk 4 above)
Comment (GuyHarwood) [reply pupil-spa/src/app/services/device/device.service.spec.ts:54]: ahh jeez, yes i'm auto thinking you are in `tslib` 🤦‍♂️ 😄

**Review by GuyHarwood — APPROVED** [review]

**Pull request merged**